### PR TITLE
Debian Improvements

### PR DESCRIPTION
**Pull request by Anuskuss**
Disclaimer: I'm not a Debian maintainer and I'm sure there are some edge cases not handled by the way I do the pre/post scripts (for example I call `systemctl` directly instead of `deb-systemd-invoke`) but this works for me. I'm sure this can be improved upon once submitted to/picked up from the Debian repo.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new configuration file structure for the `peerbanhelper` package, enhancing organization with a consolidated configuration file and a separate profile settings file.
	- Added a new dependency for user management during installation.

- **Bug Fixes**
	- Improved error handling and user creation logic in the installation script.

- **Documentation**
	- Updated service configuration to enhance logging capabilities and specify new parameters.

- **Chores**
	- Added scripts for proper handling of service management during installation and uninstallation processes.

### CODE REVIEW

**Comment by Anuskuss** [comment]
The bot is pretty cool but I think I'm going to ignore all of the suggestions. The last suggestion (security hardening) is of course useful but is generally not done for other Debian packages and would require some futher testing. Feel free to add them yourself via `systemctl edit peerbanhelper` though.

File: pkg/deb/usr/lib/systemd/system/peerbanhelper.service
Comment (Gaojianli) [inline pkg/deb/usr/lib/systemd/system/peerbanhelper.service:8]: According to the FHS specification, data files should not be stored in the `/etc` directory, I'll ask @Ghost-chu  to add an option `pbh.configdir`, let's put the data files under `/var/lib` and leave config in `/etc`

File: pkg/deb/DEBIAN/postinst
Comment (Gaojianli) [inline pkg/deb/DEBIAN/postinst:5]: Auto start? I don't think this should be written in the post-install script, this should be executed by user

**Review by Gaojianli — CHANGES_REQUESTED** [review]

File: pkg/deb/DEBIAN/postrm
Comment (paulzzh) [inline pkg/deb/DEBIAN/postrm:1]: Should the configuration file be deleted?

File: pkg/deb/usr/lib/systemd/system/peerbanhelper.service
Comment (Ghost-chu) [reply pkg/deb/usr/lib/systemd/system/peerbanhelper.service:8]: @Gaojianli Changes applied, also I think it's un-necessary to package config.yml or profile.yml in .deb, because PeerBanHelper will auto-generate them.

File: pkg/deb/DEBIAN/postrm
Comment (Gaojianli) [reply pkg/deb/DEBIAN/postrm:1]: This script is only for `apt purge`, the config should be deleted

File: pkg/deb/DEBIAN/postinst
Comment (Anuskuss) [reply pkg/deb/DEBIAN/postinst:5]: Why? PBH starts with a wizard anyway (doesn't do anything without configuration) and the user might not know that they need to `systemctl start` the service. The only thing that would suck is if 9898 is already in use making the installation not "finish successfully" but since it's the last command anyway it might not be that big of a deal.

File: pkg/deb/usr/lib/systemd/system/peerbanhelper.service
Comment (Anuskuss) [reply pkg/deb/usr/lib/systemd/system/peerbanhelper.service:8]: The config files need to be inside the DEB otherwise `conffiles` can't be used.

**Comment by Anuskuss** [comment]
I implemented the suggested changes but I haven't actually tested them because I'm still using v7.1.1. Also do you want me to research how to deal with the config changes or is it okay to break existing installations? How's the policy on that?

File: pkg/deb/usr/lib/systemd/system/peerbanhelper.service
Comment (Ghost-chu) [reply pkg/deb/usr/lib/systemd/system/peerbanhelper.service:8]: @Gaojianli I think this change should be implemented, but will require migration of existing files.

File: pkg/deb/DEBIAN/postinst
Comment (Ghost-chu) [reply pkg/deb/DEBIAN/postinst:5]: I think @Gaojianli would like us to leave it unchanged and let the end user decide and choose whether PeerBanHelper's service should be started automatically or not.
This is not a problem for professional Linux operations, but considering that PeerBanHelper users may simply install the package on their HomeLab or NAS, this action is worthwhile in my opinion.

Comment (Gaojianli) [reply pkg/deb/DEBIAN/postinst:5]: the systemd is set with `Restart=on-failure`, which means if 9898 is not work, the pbh will keep restarting and drain the resource of the system. Anyway, please remove this. For user who don't know `systemctl start`, we can add a message to notice them, just like the freebsd version: 
https://github.com/PBH-BTN/PeerBanHelper/blob/21ea4eac44763979b3f30eb9bbe8afb75aeae0b9/pkg/pkg/%2BDISPLAY#L2

File: pkg/deb/usr/lib/systemd/system/peerbanhelper.service
Comment (Gaojianli) [inline pkg/deb/usr/lib/systemd/system/peerbanhelper.service:8]: the `datadir` should be `/var/lib/peerbanhelper`, don't forget to create those dir~

File: pkg/deb/usr/lib/systemd/system/peerbanhelper.service
Comment (Gaojianli) [reply pkg/deb/usr/lib/systemd/system/peerbanhelper.service:8]: @Anuskuss  this change can be implemented

**Comment by Gaojianli** [comment]
> I implemented the suggested changes but I haven't actually tested them because I'm still using v7.1.1. Also do you want me to research how to deal with the config changes or is it okay to break existing installations? How's the policy on that?

For migration we may check existing config in the post-install script and move them. The log change will be deployed in next patch, today night or tomorrow

**Comment by Gaojianli** [comment]
> I implemented the suggested changes but I haven't actually tested them because I'm still using v7.1.1. Also do you want me to research how to deal with the config changes or is it okay to break existing installations? How's the policy on that?

For migration we may check existing config in the post-install script and move them. The log change will be deployed in next patch, today night or tormmorow

File: pkg/deb/usr/lib/systemd/system/peerbanhelper.service
Comment (Ghost-chu) [reply pkg/deb/usr/lib/systemd/system/peerbanhelper.service]: @coderabbitai stop the same suggestion in this PR

File: pkg/deb/DEBIAN/postinst
Comment (Anuskuss) [reply pkg/deb/DEBIAN/postinst:5]: > which means if 9898 is not work, the pbh will keep restarting and drain the resource of the system.

But only up to 5 times then it will enter the failed state.

> Anyway, please remove this.

What exactly? The starting of the service, the enabling (start at boot but not right now) or both?

Comment (Ghost-chu) [reply pkg/deb/DEBIAN/postinst:5]: @Anuskuss I checked this script and I noticed an issue, if the user disables the service autostart after installing the package, this script seems to re-enable the service autostart when upgrading the package.

Is it possible to enable autostart only on the first installation?   

I would like the install script to honor the user's choice. Because it is the user's right to cancel autostart. So we shouldn't override it while upgrading.

Comment (Gaojianli) [reply pkg/deb/DEBIAN/postinst:5]: > > which means if 9898 is not work, the pbh will keep restarting and drain the resource of the system.
> 
> But only up to 5 times then it will enter the failed state.
> 
> > Anyway, please remove this.
> 
> What exactly? The starting of the service, the enabling (start at boot but not right now) or both?

do nothing but 
```
echo "To start peerbanhelper, please run 'systemctl start peerbanhelper.service'"
echo "To auto-start, please run 'systemctl enable peerbanhelper.service'"
```

File: pkg/deb/usr/lib/systemd/system/peerbanhelper.service
Comment (Anuskuss) [reply pkg/deb/usr/lib/systemd/system/peerbanhelper.service:8]: I forgot to change the `WorkingDirectory` (because I didn't test it :P). Is this option actually needed though? Now that the data/config dirs are definded I'm not sure PBH actually does something in relative paths. The only other thing I can think of is setting it to `/usr/lib/peerbanhelper` so you can to `-jar PeerBanHelper.jar` (which seems rather useless).
The JAR should probably also go in `/opt` to be honest.

File: pkg/deb/usr/lib/systemd/system/peerbanhelper.service
Comment (Anuskuss) [reply pkg/deb/usr/lib/systemd/system/peerbanhelper.service:8]: I went with `/opt` instead but if you like `/var/lib` more I can change it to that.

**Comment by Ghost-chu** [comment]
> the systemd is set with Restart=on-failure, which means if 9898 is not work, the pbh will keep restarting and drain the resource of the system. Anyway, please remove this. For user who don't know systemctl start, we can add a message to notice them, just like the freebsd version:

@Gaojianli 
I think the first launch is fine, as long as the user is able to turn it off. 

I reject this suggestion: https://github.com/PBH-BTN/PeerBanHelper/pull/698#discussion_r1832184160

File: pkg/deb/usr/lib/systemd/system/peerbanhelper.service
Comment (Gaojianli) [reply pkg/deb/usr/lib/systemd/system/peerbanhelper.service:8]: > I forgot to change the `WorkingDirectory` (because I didn't test it :P). Is this option actually needed though? Now that the data/config dirs are definded I'm not sure PBH actually does something in relative paths. The only other thing I can think of is setting it to `/usr/lib/peerbanhelper` so you can to `-jar PeerBanHelper.jar` (which seems rather useless). The JAR should probably also go in `/opt` to be honest.

I am not sure where the JAR should be placed... I googled it and someone tell me I should put it in the `/usr/lib/`. But for /opt, I think this directory is more likely used for compiled software. If I am wrong, please feel free to point it.

Comment (Ghost-chu) [reply pkg/deb/usr/lib/systemd/system/peerbanhelper.service:8]: > I forgot to change the `WorkingDirectory` (because I didn't test it :P). Is this option actually needed though? Now that the data/config dirs are definded I'm not sure PBH actually does something in relative paths. The only other thing I can think of is setting it to `/usr/lib/peerbanhelper` so you can to `-jar PeerBanHelper.jar` (which seems rather useless). The JAR should probably also go in `/opt` to be honest.

You are free to set it up and change it, and if it involves a location that PBH doesn't currently support modifying, I can modify the code to accommodate the startup command line.

**Comment by Gaojianli** [comment]
> @Gaojianli I think the first launch is fine, as long as the user is able to turn it off.
> 
> I reject this suggestion: [#698 (comment)](https://github.com/PBH-BTN/PeerBanHelper/pull/698#discussion_r1832184160)

Well, `systemctl restart peerbanhelper.service` is accepted for me

File: pkg/deb/DEBIAN/postinst
Comment (Anuskuss) [reply pkg/deb/DEBIAN/postinst:5]: > Is it possible to enable autostart only on the first installation?

Yeah sure you can filter by event so on `install` you enable and on `upgrade` you do nothing.

> do nothing

Sure I can do that. Hopefully the user reads this though but I guess Linux users really don't need much hand-holding. Will implement this once I get back from work.

**Comment by Ghost-chu** [comment]
> > @Gaojianli I think the first launch is fine, as long as the user is able to turn it off.
> > I reject this suggestion: [#698 (comment)](https://github.com/PBH-BTN/PeerBanHelper/pull/698#discussion_r1832184160)
> 
> Well, `systemctl restart peerbanhelper.service` is accepted for me



> > @Gaojianli I think the first launch is fine, as long as the user is able to turn it off.
> > I reject this suggestion: [#698 (comment)](https://github.com/PBH-BTN/PeerBanHelper/pull/698#discussion_r1832184160)
> 
> Well, `systemctl restart peerbanhelper.service` is accepted for me

@Gaojianli 
I think autostart is great, geez let's go to IM and argue and not keep sending tons of emails.

File: pkg/deb/DEBIAN/postinst
Comment (Ghost-chu) [reply pkg/deb/DEBIAN/postinst:5]: > > Is it possible to enable autostart only on the first installation?
> 
> Yeah sure you can filter by event so on `install` you enable and on `upgrade` you do nothing.
> 
> > do nothing
> 
> Sure I can do that. Hopefully the user reads this though but I guess Linux users really don't need much hand-holding. Will implement this once I get back from work.

We are discussing the changes regarding autostart and will get in touch with you when we have new results.

**Comment by Gaojianli** [comment]
> > > @Gaojianli I think the first launch is fine, as long as the user is able to turn it off.
> > > I reject this suggestion: [#698 (comment)](https://github.com/PBH-BTN/PeerBanHelper/pull/698#discussion_r1832184160)
> > 
> > 
> > Well, `systemctl restart peerbanhelper.service` is accepted for me
> 
> > > @Gaojianli I think the first launch is fine, as long as the user is able to turn it off.
> > > I reject this suggestion: [#698 (comment)](https://github.com/PBH-BTN/PeerBanHelper/pull/698#discussion_r1832184160)
> > 
> > 
> > Well, `systemctl restart peerbanhelper.service` is accepted for me
> 
> @Gaojianli I think autostart is great, geez let's go to IM and argue and not keep sending tons of emails.

```bash
SERVICE="peerbanhelper.service"
INSTALLED_FLAG="/etc/peerbanhelper/.installed"
if [ -e $INSTALLED_FLAG ]; then
    # update
    if systemctl is-active --quiet "$SERVICE" > /dev/null 2>&1; then
        echo "$SERVICE is already running. Restarting the service."
        systemctl restart "$SERVICE"
    fi
else
    # first install
    systemctl enable --now $SERVICE
    touch $INSTALLED_FLAG
fi
```

File: pkg/deb/DEBIAN/postinst
Comment (Gaojianli) [reply pkg/deb/DEBIAN/postinst:5]: > > Is it possible to enable autostart only on the first installation?
> 
> Yeah sure you can filter by event so on `install` you enable and on `upgrade` you do nothing.
> 
> > do nothing
> 
> Sure I can do that. Hopefully the user reads this though but I guess Linux users really don't need much hand-holding. Will implement this once I get back from work.

can you ref this script?
```bash
SERVICE="peerbanhelper.service"
INSTALLED_FLAG="/etc/peerbanhelper/.installed"
if [ -e $INSTALLED_FLAG ]; then
    # update
    if systemctl is-active --quiet "$SERVICE" > /dev/null 2>&1; then
        echo "$SERVICE is already running. Restarting the service."
        systemctl restart "$SERVICE"
    fi
else
    # first install
    systemctl enable --now $SERVICE
    touch $INSTALLED_FLAG
fi
```

**Comment by Anuskuss** [comment]
I fixed some of the things you mentioned:
- There's an upgrade path now which puts the old layout into their correct location (hopefully I didn't screw this up)
- I've realize now that it doesn't even make sense to autostart PBH because if the user doesn't even autostart their torrent client PBH will just waste resources. So the user has to enable it themselves (perferrably with something like `BindsTo=transmission-daemon.service`).
- The service will still start automatically because I think that's what the user excepts after installing it. I can also start it only if it was active before upgrading but that would require some state tracking as the service gets stopped in `prerm` and I need to communicate that to `postinst`.

Any other suggestions?

File: pkg/deb/DEBIAN/postinst
Comment (Anuskuss) [reply pkg/deb/DEBIAN/postinst]: Autostart is gone now.

File: pkg/deb/usr/lib/systemd/system/peerbanhelper.service
Comment (Anuskuss) [reply pkg/deb/usr/lib/systemd/system/peerbanhelper.service]: It's in `/var/lib` now.

**Comment by Anuskuss** [comment]
Hm there's a `java.io.FileNotFoundException: data/logs/latest.log (No such file or directory)` error now. Can you drop that code completely? The `data` directory contains duplicate YAML configs and a duplicate log so I figured I'd just delete that directory completely.

**Comment by Ghost-chu** [comment]
> Hm there's a error now. Can you drop that code completely? The directory contains duplicate YAML configs and a duplicate log so I figured I'd just delete that directory completely.`java.io.FileNotFoundException: data/logs/latest.log (No such file or directory)``data`

Could you provide a copy of the exception stacktrace?

**Comment by Anuskuss** [comment]
Just delete the `data` directory and see for yourself. Because the working directory is now `/usr/lib/peerbanhelper` (no write access) it can't create it. But even if I changed it to `/var/lib/peerbanhelper` what's the point? It will literally just create a `logs` directory in there with an empty `latest.log`. Just remove that code completely.

<details><summary>Log</summary>

```
java[627895]: |-ERROR in ch.qos.logback.core.rolling.RollingFileAppender[FILE] - Failed to create parent directories for [/usr/lib/peerbanhelper/data/logs/latest.log]
java[627895]: |-ERROR in ch.qos.logback.core.rolling.RollingFileAppender[FILE] - openFile(data/logs/latest.log,true) call failed. java.io.FileNotFoundException: data/logs/latest.log (No such file or directory)
java[627895]:         at java.io.FileNotFoundException: data/logs/latest.log (No such file or directory)
java[627895]:         at         at java.base/java.io.FileOutputStream.open0(Native Method)
java[627895]:         at         at java.base/java.io.FileOutputStream.open(Unknown Source)
java[627895]:         at         at java.base/java.io.FileOutputStream.<init>(Unknown Source)
java[627895]:         at         at ch.qos.logback.core.recovery.ResilientFileOutputStream.<init>(ResilientFileOutputStream.java:26)
java[627895]:         at         at ch.qos.logback.core.FileAppender.openFile(FileAppender.java:206)
java[627895]:         at         at ch.qos.logback.core.FileAppender.start(FileAppender.java:126)
java[627895]:         at         at ch.qos.logback.core.rolling.RollingFileAppender.start(RollingFileAppender.java:104)
java[627895]:         at         at ch.qos.logback.core.model.processor.AppenderModelHandler.postHandle(AppenderModelHandler.java:84)
java[627895]:         at         at ch.qos.logback.core.model.processor.DefaultProcessor.secondPhaseTraverse(DefaultProcessor.java:257)
java[627895]:         at         at ch.qos.logback.core.model.processor.DefaultProcessor.secondPhaseTraverse(DefaultProcessor.java:253)
java[627895]:         at         at ch.qos.logback.core.model.processor.DefaultProcessor.traversalLoop(DefaultProcessor.java:90)
java[627895]:         at         at ch.qos.logback.core.model.processor.DefaultProcessor.process(DefaultProcessor.java:106)
java[627895]:         at         at ch.qos.logback.core.joran.GenericXMLConfigurator.processModel(GenericXMLConfigurator.java:222)
java[627895]:         at         at ch.qos.logback.core.joran.GenericXMLConfigurator.doConfigure(GenericXMLConfigurator.java:178)
java[627895]:         at         at ch.qos.logback.core.joran.GenericXMLConfigurator.doConfigure(GenericXMLConfigurator.java:123)
java[627895]:         at         at ch.qos.logback.core.joran.GenericXMLConfigurator.doConfigure(GenericXMLConfigurator.java:66)
java[627895]:         at         at ch.qos.logback.classic.util.DefaultJoranConfigurator.configureByResource(DefaultJoranConfigurator.java:68)
java[627895]:         at         at ch.qos.logback.classic.util.DefaultJoranConfigurator.configure(DefaultJoranConfigurator.java:35)
java[627895]:         at         at ch.qos.logback.classic.util.ContextInitializer.invokeConfigure(ContextInitializer.java:128)
java[627895]:         at         at ch.qos.logback.classic.util.ContextInitializer.autoConfig(ContextInitializer.java:103)
java[627895]:         at         at ch.qos.logback.classic.util.ContextInitializer.autoConfig(ContextInitializer.java:66)
java[627895]:         at         at ch.qos.logback.classic.spi.LogbackServiceProvider.initializeLoggerContext(LogbackServiceProvider.java:52)
java[627895]:         at         at ch.qos.logback.classic.spi.LogbackServiceProvider.initialize(LogbackServiceProvider.java:41)
java[627895]:         at         at org.slf4j.LoggerFactory.bind(LoggerFactory.java:199)
java[627895]:         at         at org.slf4j.LoggerFactory.performInitialization(LoggerFactory.java:186)
java[627895]:         at         at org.slf4j.LoggerFactory.getProvider(LoggerFactory.java:496)
java[627895]:         at         at org.slf4j.LoggerFactory.getILoggerFactory(LoggerFactory.java:482)
java[627895]:         at         at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:431)
java[627895]:         at         at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:457)
java[627895]:         at         at com.ghostchu.peerbanhelper.Main.<clinit>(Main.java:43)
java[627895]:         at         at com.ghostchu.peerbanhelper.MainJumpLoader.main(MainJumpLoader.java:17)
```
</details>

I'm also thinking about adding `SuccessExitStatus=143` (I'm already using that locally) because the unit always returns that exit code for seemingly no reason. Unless you find the root cause I think it wouldn't hurt.

**Review by Ghost-chu — APPROVED** [review]
PR looks good and ready for merge.

How ever I need take a look into that exception but I don't have much time in recent days so I guess I will get back at 25-30/11 around.

**Comment by Anuskuss** [comment]
> PR looks good and ready for merge.

Have you actually tested this? Because for the migration path I only did tests with minimal configuration and I don't know if I missed something. After running it for a couple of days now I see that there are `.log.gz` files in `logs` and I previously only copied `latest.log` (now I even delete the directory completely) and that definitely needs to be fixed. As for the normal install/upgrade I only did small changes anyway so I think that's good to go.

> I need take a look into that exception

I still want to remove `data` so I guess the user has to live with that exception for now. Doesn't seem to be that serious anyway.

**Comment by Gaojianli** [comment]
LGTM, once you rebase the target branch, this pr can be merged

**Comment by Anuskuss** [comment]
Resubmitted as #732.

**Pull request closed without merging**
